doc/architecture/decisions/0015-key-management.md

```diff
@@ -0,0 +1,222 @@
+
+To date, there we have done a bunch of work on DPKI, Personas, and hierarchical deterministic keys documented here:
+
+ - Usage Story: https://hackmd.io/87H6J0pZSaG8MgbpnbjlIA
+ - Long ADR draft: https://hackmd.io/VIVvRD9gQ5KnDy_C-y6leA
+ - David's thoughts on DPKI: https://hackmd.io/2rDYowopS9OwRazptpQMYA
+
+These documents detail various uses cases, including initial sign-up, key revocation, etc.  The central insight from this work is the need to create a Hierarchical Deterministic Key generation system, based on a Primary Seed, from which additional seeds can be generated which then are in turn used to actually generate many key-pairs.  This allows us, by-convention, to use the first seed generated by the Primary seed as the seed for revocation keys, and subsequent seeds as seeds for keys of separate Holochain devices that can be proven to be under the control of the holder of Primary Seed.
+
+This ADR documents the synthesis of those documents in the form of functions that the Holochain core library needs to make available to container/composer developers as well as the app API needs to make available to DPKI holochain app developers via special capabilities.
+
+These functions must provide the ability for associating key-pairs with a DPKI identity such that the private key never leaves the device it was generated on.
+
+Libraries and standards exist for creating seeds and mnemonic word lists for those seeds.  These standards are in use with hardware wallets, that we may want Holo and Holochain to be able to interoperate with. Though there are some questions as to what degree this will be feasible given the rate and frequency that various Holochain processes need signing to happen which would make it impractical to keep having to do click on a hardware wallet.  I.e. any networking operation signing key cannot feasibly live on Trezor/Ledger.
+
+Secp256k1 wasn’t designed for encryption, but for signing, so at least some of our seeds and keys need to be generated using other standards.
+
+### Illustrative Workflows
+
+##### 1 - new / first holoport
+
+```rust
+// generate new random seed hierarchy
+let hd = DpkiHd::new_random();
+
+// display the primary/root seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_primary_seed()));
+
+// publish to the dpki dht
+let app_pub_data = DpkiApp::new_defaults(hd);
+publish_to_dpki_dht(app_pub_data);
+
+// initialize a device structure with device seed number 1
+let device = DpkiDevice::new_from_seed(hd.get_device_seed(1));
+// (optional) display device seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_device_seed(1)));
+// blink the device pin on the holoport circle
```

Comment:
?

Reply:
I think David's idea here is that that would be a way to display a pin by blinking it on the Holo-port circle, but I don't think that's serious...

Reply:
It is serious. Following the principle that private keys (and seeds for key pairs) should never be sent over the wire/network, how do you propose to "display" keys from a HoloPort which people do not have a monitor plugged into? 

We are targeting HoloPorts as turnkey plug/play devices for non-techies. Non-techies often do not have spare keyboards and monitors laying around their house anymore. Most people primarily use consumer grade devices like phones/tablets/laptops now. If we want to give people a way to record critical information to regenerate their system, we can use the Holo LED to flash numbers to them, changing colors for each next number, so they can write down codes to regenerate keys.

Sending it to a phone or computer over the network goes against basic rules of private key management. It COULD be done if people want to take that risk, but we should allow people the more secure option to have codes flashed to them.

Reply:
Three words: Integrated mini printer. 😂

Reply:
it is a "holo"-port... where's the holographic projector?

Reply:
can we mimic how ledger approaches initialization of seeds and pins?